server/src/main/java/org/elasticsearch/search/fetch/subphase/StoredFieldsPhase.java

```diff
@@ -68,21 +56,19 @@ public FetchSubPhaseProcessor getProcessor(FetchContext fetchContext) {
         }
 
         // build the StoredFieldsSpec and a list of StoredField records to process
-        List<StoredField> storedFields = new ArrayList<>(METADATA_FIELDS);
+        List<StoredField> storedFields = new ArrayList<>();
         Set<String> fieldsToLoad = new HashSet<>();
         if (storedFieldsContext.fieldNames() != null) {
             SearchExecutionContext sec = fetchContext.getSearchExecutionContext();
             for (String field : storedFieldsContext.fieldNames()) {
-                if (SourceFieldMapper.NAME.equals(field) == false) {
-                    Collection<String> fieldNames = sec.getMatchingFieldNames(field);
-                    for (String fieldName : fieldNames) {
-                        MappedFieldType ft = sec.getFieldType(fieldName);
-                        if (ft.isStored() == false) {
-                            continue;
-                        }
-                        storedFields.add(new StoredField(fieldName, ft, sec.isMetadataField(ft.name())));
-                        fieldsToLoad.add(ft.name());
+                Collection<String> fieldNames = sec.getMatchingFieldNames(field);
+                for (String fieldName : fieldNames) {
```

Comment:
The way it is written now means that this phase `StoredFieldPhase` is responsible for fetching stored fields (fields that have stored: true in the mapper) which are requested via `stored_fields` including when they are requested explicitly like in `stored_fields: [_one, _two]` or using `stored_fields: [ * ]`.

Stored fields which are requested via `fields` **must** be handled by FetchFieldsPhase.

Reply:
This is the way the split should be done according to my udnerstanding:
1. stored fields requested via fields => FetchFieldsPhase using ValueFetcher
2. stored fields requested via stored_fields => **StoredFieldsPhase**
3. non-stored fields requested via fields +> FetchFieldsPhase using ValueFecther
4. non-stored fields requested via stored_fields => FetchFieldsPhase using ValueFetcher
5. stored metadata fields requested via fields => FetchFieldsPhase using ValueFetcher
6. stored metadata fields requested via stored_fields => **StoredFieldsPhase**
7. non-stored metadata fields requested via fields => FetchFieldsPhase using ValueFetcher
8. non-stored metadata fields requested via stored_fields FetchFieldsPhase using ValueFetcher

As you can see this is why I have written it that way and the way that makes `SizeMappingIT` work.

Reply:
Also notice that this is the way that allows us to use the ValueFecther for **ALMOST** averything.

Reply:
Also, as a result of removing `isMetadataField` the distinction on which phase to use does not depend anymore on whether the field is a metadata field or not but just on:
* field is stored or not in the mapper (`stored: true` or `stored: false`)
* field is requested via `fields` or `stored_fields`.

Reply:
> I think we need to agree on that because at the end the way I see it is that, at least from my perspective, it is not that important which phase is responsible for what as long as:

It is important once you move `_ignored` to have doc_values instead of being stored, it needs to go via value fetchers at all times, it can't be retrieved some times via `FetchFieldsPhase` and some other times via `StoredFieldsPhase`. I explained that in another comment above.


> Anyway I think the original idea behind the refactoring is: "we want to use ValueFetcher whenever possible" which means basically for all fields that have a ValueFetcher considering that some fields do not have it (they throw UnsupportedOperationException like for instance _seq_no).

That requires special treatment, indeed. The current distinction can be made using `MappedFieldType#isStored` , but once you have `_ignored` no longer stored, we need a different mechanism. We do need to exclude fields that have never been retrieved before. The reason is now that they are not stored, but that condition will need to be adapted. I tried different approaches around this: 
1) give those fields (e.g. _seq_no) a value fetcher that returns their data: we would make them retrievable via runtime fields as well which I am not sure is what we want.
2) give those fields an empty value fetcher, there are metadata fields that are currently excluded that already have a non empty value fetcher, hence they can be retrieved via script but are excluded from the `StoredFieldsPhase`
3) add a specific method to MappedFieldType: #107086 . I think this is probably the best approach. We can keep for now the ft.isStored check, and replace it later with a different conditional. This is not needed to get the current PR in.

Reply:
>  I think there should be tests that fail because of this, or maybe tests are missing.

let me clarify this too: I was under the impression that if `StoredFieldsPhase` still retrieves metadata fields, we must keep the isMetadataField flag in the StoredField record, or those metadata fields will be added among document fields and not meta fields in the response. I do see that SearchHit has a strange way of retrieving fields via get, that pulls from both doc as well as metadata fields (as a fallback). Maybe that covers issues. IN the output, I think that doc fields are serialized under `fields` in the response, while metadata fields at the hit, at the same level as `_source`. This distinction is not so clear when checking SearchHits via IT tests.